docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+
+- IPAM providers should not be added to CAPI directly, but live as individual projects instead.
+- IPAddress and IPAddressPool will only have limited validation (e.g. that the referenced Pool exists and that IP address strings are valid addresses). Advanced validation (whether an IP is actually part of the correct subnet for example) needs to be handled by IPAM providers.
+
+## Proposal
+
+The chosen approach was heavily inspired by metal3's ip-address-manager, which in turn seems to have been inspired by Kubernetes' PersistentVolumes. The proposal is twofold and consists of the API contract that should be added to CAPI and two examples to better illustrate how the contract is used by IPAM and infrastructure providers.
+
+### IPAM API Contract
+
+The IPAM API contract provides a generic interface between CAPI infrastructure providers and to-be-developed IPAM providers. Any intersted infrastructure is supposed to be able to request IP Addresses from a defined pool of addresses. IPAM providers should then be able to find and fulfil such requests using any custom logic to do so.
+
+The contract consists of three resources:
+
+- An IPAddressClaim, which is used to allocate addresses
+- An IPAddress that IPAM providers use to fulfil IPAddressClaims
+- IP Pools, which hold provider specific configuration on how addresses should be allocated, e.g. a specific subnet to use
+
+Both **IPAddressClaims** and **IPAddresses** should be part of Cluster API, while **IPPools** are defined by the different IPAM providers.
+
+An **IPAddressClaim** is used by infrastructure providers to request an IP Address. The claim contains a reference to an IP Pool. Because the pool is provider specific, the IPAM controllers can decide whether to handle a claim by inspecting the group and kind of the pool reference.
+
+If a IPAM controller detects a Claim that references a Pool it controls, it allocates an IP address from that pool and creates an **IPAddress** to fulfil the claim. It also updates the status of the IPAddressClaim with a reference to the created Address.
+
+```go
+// IPAddressClaimSpec describes the desired state of an IPAddressClaim
+type IPAddressClaimSpec struct {
+   // Pool is a reference to the pool from which an IP address should be allocated.
+   Pool LocalObjectReference `json:"pool,omitempty"`
+}
+
+// IPAddressClaimStatus contains the status of an IPAddressClaim
+type IPAddressClaimStatus struct {
+   // Address is a reference to the address that was allocated for this claim.
+   Address LocalObjectReference `json:"address,omitempty"`
+}
+ 
+// IPAddressClaim can be used to allocate IPAddresses from an IP Pool.
+type IPAddressClaim struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec   IPAddressClaimSpec   `json:"spec,omitempty"`
+   Status IPAddressClaimStatus `json:"status,omitempty"`
+}
+```
+
+```go
+// IPAddressSpec describes an IPAddress
+type IPAddressSpec struct {
```

Comment:
I think IPAddresses should be immutable. It's infrastructure provider specific whether it's possible to change the IP Address of a machine after it's been created, which could then make it hard/impossible to implement this contract. Since machines are treated as cattle, rolling the machine to get a new address should be sufficient imo.

I guess that depends on how providers implement the proposal on their side. The infra Machine won't contain a reference to the IPAddressClaim, as the Machine is created by CAPI which doesn't handle that. So the claim is created by the infra provider, and linked to the Machine using e.g. the name.
Most providers allow to specify IPs already, so I think that mechanism should just continue to exist so people can specify addresses manually as well.

Reply:
Thanks for clarifying. I think it will be useful if some of this can be captured in the doc as a recommendation for the implementers, but I leave the final call on this to you